[AWS Cloudwatch] Fix Tags for APIGW

gizas opened the pull request:
- Enhancement

## Proposed commit message


- WHAT: Enhance cloudwatch metricset when APIGW is used as a namespace to be able to retrieve the tags configured in AWS 
- WHY: Currently the aws.tags retrieval was not working because we had no way to correlate tags of RestAPis with provided metrics of cloudwatch  


## Checklist


- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.


## How to test this PR locally

1. Install metricbeat with current configuration:
```yaml
    - module: aws
      period: 10s
      regions:
        - us-east-1
      metricsets:
        - cloudwatch
      metrics:
        - namespace: "AWS/ApiGateway"
          resource_type: "apigateway:restapis"
          tags_filter: 
             - key: "ServiceOwner"
              value: "ole"
     
```
2. Create RestAPi in AWS and configure some tags eg. `ServiceOwner:ole`

## Related issues


- Closes https://github.com/elastic/beats/issues/33913

## Use cases

See analysis https://github.com/elastic/sdh-beats/issues/5103#issuecomment-2333853534

## Screenshots

Successful usage 
![Screenshot 2024-09-11 at 5 21 24 PM](https://github.com/user-attachments/assets/579b5e1f-f55b-4861-aa38-11bb2c36052f)


## Logs

```bash
.version":"1.6.0"}}
{"log.level":"info","@timestamp":"2024-09-11T14:22:12.226Z","log.logger":"aws.cloudwatch","log.origin":{"function":"github.com/elastic/beats/v7/x-pack/metricbeat/module/aws/cloudwatch.(*MetricSet).Fetch","file.name":"cloudwatch/cloudwatch.go","file.line":210},"message":"PASSS InfoAPI response: map[PetStore:t8diprvp81 apm-typescript-test-dev:v7o6rbtm6i gizastest2:0yqjzpf5c7 gizastest3:w6lw8qolsl gizastest_private:x4fe7jy7o1 lucian-REST-PetStore:rsad95n9p5 testJavaApi:fooiifd0q6 tommyers-rest:p83fmmae0c]","service.name":"metricbeat","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2024-09-11T14:22:13.347Z","log.logger":"cloudwatch","log.origin":{"function":"github.com/elastic/beats/v7/x-pack/metricbeat/module/aws/cloudwatch.(*MetricSet).createEvents","file.name":"cloudwatch/cloudwatch.go","file.line":546},"message":"In region us-east-1, service /restapis/0yqjzpf5c7 tags match tags_filter","service.name":"metricbeat","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2024-09-11T14:22:13.347Z","log.logger":"cloudwatch","log.origin":{"function":"github.com/elastic/beats/v7/x-pack/metricbeat/module/aws/cloudwatch.(*MetricSet).createEvents","file.name":"cloudwatch/cloudwatch.go","file.line":546},"message":"In region us-east-1, service 0yqjzpf5c7 tags match tags_filter","service.name":"metricbeat","ecs.version":"1.6.0"}
```

constanca-m commented on x-pack/metricbeat/module/aws/cloudwatch/cloudwatch.go:
```diff
@@ -566,6 +582,15 @@ func (m *MetricSet) createEvents(svcCloudwatch cloudwatch.GetMetricDataAPIClient
 				// And tags are only store under s3BucketName in resourceTagMap.
 				subIdentifiers := strings.Split(identifierValue, dimensionSeparator)
 				for _, subIdentifier := range subIdentifiers {
+					m.logger.Infof("PASSSS Before %s", subIdentifier)
```
Extra logs?

elasticmachine commented:
Pinging @elastic/obs-ds-hosted-services (Team:obs-ds-hosted-services)

kaiyan-sheng commented on x-pack/metricbeat/module/aws/utils.go:
```diff
@@ -92,6 +96,52 @@ func GetListMetricsOutput(namespace string, regionName string, period time.Durat
+// GetRestAPIsOutput function gets results from apigw api.
+// GetRestApis Apigateway API is used to retrieve only the REST API specified info. This returns a map with the names and ids of RestAPIs configured
+// Limit variable defines maximum number of returned results per page. The default value is 25 and the maximum value is 500.
+func GetRestAPIsOutput(svcRestApi *apigateway.Client, limit *int32) (map[string]string, error) {
```
Maybe change to a name that reflects apigateway here?
```suggestion
func GetAPIGatewayAPIOutput(svcRestApi *apigateway.Client, limit *int32) (map[string]string, error) {
```

kaiyan-sheng commented on x-pack/metricbeat/module/aws/utils.go:
```diff
@@ -92,6 +96,52 @@ func GetListMetricsOutput(namespace string, regionName string, period time.Durat
+
+// GetAPIsOutput function gets results from apigatewayv2 api.
+// GetApis Apigateway API is used to retrieve the HTTP and WEBSOCKET specified info. This returns a map with the names and ids of relevant APIs configured
+func GetAPIsOutput(svcHttpApi *apigatewayv2.Client) (map[string]string, error) {
```
same here?

gizas replied:
Replaced to GetAPIGatewayRestAPIOutput and GetAPIGatewayAPIOutput

kaiyan-sheng commented on x-pack/metricbeat/module/aws/cloudwatch/cloudwatch_test.go:
```diff
@@ -1406,7 +1420,7 @@ func TestCreateEventsWithTagsFilter(t *testing.T) {
 	}
 
-	events, err = m.createEvents(mockCloudwatchSvc, mockTaggingSvc, listMetricWithStatsTotal, resourceTypeTagFilters, regionName, startTime, endTime)
+	events, err = m.createEvents(mockCloudwatchSvc, mockTaggingSvc, listMetricWithStatsTotal, resourceTypeTagFilters, infoAPImap, regionName, startTime, endTime)
```
I wonder if we can store infoAPImap result into resourceTypeTagFilters since it is collecting tags but just with a diff api. This way, we dont have to change the createEvents function input to have a special parameter for apigatewayapi.

gizas replied:
The infoAPImap has [Name_of_IP : ID_of_API] not actual tags. 
The diff:  resourceTypeTagFilters map[string][]aws.Tag and  infoAPImap map[string]string,
(I will add it somewhere in a note to make clear)

TBH we can create a new createEventsforAPIGW that will have the new infoAPImap but too many same code lines just for one argument. WDYT?

kaiyan-sheng replied:
Yeah lets keep it as is then, thanks!!

gizas commented on x-pack/metricbeat/module/aws/cloudwatch/cloudwatch.go:
```diff
@@ -189,14 +192,49 @@ func (m *MetricSet) Fetch(report mb.ReporterV2) error {
+						m.Logger().Errorf("could not get http and websocket apis output: %v", err)
+					}
+					if len(infotherapi) > 0 {
+						maps.Copy(infoapi, infotherapi)
```
There is this lint error which I think can be ignored. 
The alternative can be sth like:
```
func mergeMaps(m1 map[string]string, m2 map[string]string) {
	for k, v := range m2 {
		m1[k] = v
	}
}
```

WDYT?

Kavindu-Dodan replied:
seems this is the first time `maps.Copy` used in beats codebase :D that said, I think it could be due to an outdated linter. I think ignore should be fine

Kavindu-Dodan commented on x-pack/metricbeat/module/aws/cloudwatch/cloudwatch.go:
```diff
@@ -189,14 +192,49 @@ func (m *MetricSet) Fetch(report mb.ReporterV2) error {
 
-			eventsWithIdentifier, err := m.createEvents(svcCloudwatch, svcResourceAPI, filteredMetricWithStatsTotal, resourceTypeTagFilters, regionName, startTime, endTime)
+			//Check whether namespace is APIGW
+			checkns := "AWS/ApiGateway"
+			checkresource_type := "apigateway:restapis"
```
nit - I think these should be constants as they are AWS constants. And we could derive their lowercase var to avoid lower string generation in this for loop :)

gizas replied:
Made them as const in https://github.com/orgs/elastic/projects/1228/views/18

Kavindu-Dodan replied:
Nice :) 

Further, we could derive their lowercase var so we don't have to use  for example `strings.ToLower(checkresource_type)` in loops (which is as costly as ToLower generate a new string)

gizas replied:
Done in [7bd7324](https://github.com/elastic/beats/pull/40755/commits/7bd732430804135067685728fbf0e669f4abd278)

Kavindu-Dodan commented on x-pack/metricbeat/module/aws/cloudwatch/cloudwatch.go:
```diff
@@ -219,7 +257,7 @@ func (m *MetricSet) Fetch(report mb.ReporterV2) error {
 
 // createAwsRequiredClients will return the two necessary client instances to do Metric requests to the AWS API
-func (m *MetricSet) createAwsRequiredClients(beatsConfig awssdk.Config, regionName string, config aws.Config) (*cloudwatch.Client, *resourcegroupstaggingapi.Client, error) {
+func (m *MetricSet) createAwsRequiredClients(beatsConfig awssdk.Config, regionName string, config aws.Config) (*cloudwatch.Client, *resourcegroupstaggingapi.Client, *apigateway.Client, *apigatewayv2.Client, error) {
```
nit - should we create a return struct here instead of adding more return values ?

gizas replied:
Changed it to struct in  https://github.com/orgs/elastic/projects/1228/views/18
I liked the idea. I might need to test some other namespace maybe to see if we break sth? Integration tests pass

Kavindu-Dodan replied:
Nice :) . `createAwsRequiredClients` is not exported so shouldn't cause any issues :)

kaiyan-sheng commented on x-pack/metricbeat/module/aws/aws.go:
```diff
@@ -36,6 +36,7 @@ type Config struct {
 	AWSConfig             awscommon.ConfigAWS `config:",inline"`
 	TagsFilter            []Tag               `config:"tags_filter"`
 	IncludeLinkedAccounts *bool               `config:"include_linked_accounts"`
+	LimitRestAPI          *int32              `config:"limit_rest_api"`
```
Can we add some documentatino on what's this new config parameter LimitRestAPI please? Thanks!!

gizas replied:
Sure , I added [this](https://github.com/elastic/beats/pull/40755/files#diff-a43aae747ff228e3bf61ee6f27e2898587db11b4ae37286304ecb6d58181ceb7) and a comment [here](https://github.com/elastic/beats/pull/40755/files#diff-bafb9a6c22e4a9b94bfb65329d4cf4035b0664e5781192f2ac8a27f64a2c843dR109)

kaiyan-sheng replied:
Thanks!! Now I wonder if we should even expose this parameter to the users. Can we just use the default or the maximum value hardcoded instead? 
@zmoog WDYT?

Kavindu-Dodan replied:
If we plan to keep this option, then we should mention the accepted value range and may be even validate the input to match with restrictions (ex:- max 500)

gizas replied:
Added a check here https://github.com/elastic/beats/pull/40755/commits/26a4375ba45c4cb8f35480c75cf8e6afb342637c

kaiyan-sheng commented on x-pack/metricbeat/module/aws/cloudwatch/cloudwatch_test.go:
```diff
@@ -1255,13 +1257,23 @@ func (m *MockResourceGroupsTaggingClient) GetResources(context.Context, *resourc
+type MockResourceGroupsTaggingClient2 struct{}
+
+// GetResources implements resourcegroupstaggingapi.GetResourcesAPIClient.
+func (m *MockResourceGroupsTaggingClient2) GetResources(context.Context, *apigateway.GetResourcesInput, ...func(*apigateway.Options)) (*restapi.GetResourcesOutput, error) {
```
```suggestion
func (m *MockResourceGroupsTaggingClient2) GetResources(context.Context, *apigateway.GetResourcesInput, ...func(*apigateway.Options)) (*apigateway.GetResourcesOutput, error) {
	return &apigateway.GetResourcesOutput{}, nil
}
```

kaiyan-sheng commented on x-pack/metricbeat/module/aws/cloudwatch/cloudwatch_test.go:
```diff
@@ -13,6 +13,8 @@ import (
 	"time"
 
+	"github.com/aws/aws-sdk-go-v2/service/apigateway"
+	restapi "github.com/aws/aws-sdk-go-v2/service/apigateway"
```
We can remove `restapi "github.com/aws/aws-sdk-go-v2/service/apigateway"` and just use "github.com/aws/aws-sdk-go-v2/service/apigateway"

gizas replied:
Was a leftover, indeed.
Removed

Kavindu-Dodan commented on x-pack/metricbeat/module/aws/cloudwatch/cloudwatch.go:
```diff
@@ -189,14 +202,51 @@ func (m *MetricSet) Fetch(report mb.ReporterV2) error {
-			eventsWithIdentifier, err := m.createEvents(svcCloudwatch, svcResourceAPI, filteredMetricWithStatsTotal, resourceTypeTagFilters, regionName, startTime, endTime)
+			//Check whether namespace is APIGW
+			var (
+				checkns_lower            = strings.ToLower(checkns)
+				checkresource_type_lower = strings.ToLower(checkresource_type)
```
These can be moved to var definitions at line 29 of this file :) This avoids unnecessary var creation inside the loop that starts at line 203

gizas replied:
yes done in [535b905](https://github.com/elastic/beats/pull/40755/commits/535b9059fa11776e454bfd89742c7e9cdd78b555)

kaiyan-sheng commented on x-pack/metricbeat/module/aws/cloudwatch/cloudwatch.go:
```diff
@@ -189,14 +202,51 @@ func (m *MetricSet) Fetch(report mb.ReporterV2) error {
+					}
+				}
+
+				m.Logger().Infof("infoapi response: %v", infoapi)
```
should this be info level or debug level?

gizas replied:
Yes done in [535b905](https://github.com/elastic/beats/pull/40755/commits/535b9059fa11776e454bfd89742c7e9cdd78b555)

zmoog commented on x-pack/metricbeat/module/aws/cloudwatch/cloudwatch.go:
```diff
@@ -191,14 +206,50 @@ func (m *MetricSet) Fetch(report mb.ReporterV2) error {
+					}
+				}
+			}
+			if strings.Contains(strings.ToLower(namespace), checkns_lower) {
+				// inforestapi includes only Rest APIs
+				if useonlyrest {
+					infoapi, err = aws.GetAPIGatewayRestAPIOutput(APIClients.Apigateway, config.LimitRestAPI)
+					if err != nil {
+						m.Logger().Errorf("could not get rest apis output: %v", err)
+					}
+				} else {
+					// infoapi  includes only Rest APIs
+					// apiGatewayAPI includes only WebSocket and HTTP APIs
+					infoapi, err = aws.GetAPIGatewayRestAPIOutput(APIClients.Apigateway, config.LimitRestAPI)
+					if err != nil {
+						m.Logger().Errorf("could not get rest apis output: %v", err)
+					}
+
+					apiGatewayAPI, err := aws.GetAPIGatewayAPIOutput(APIClients.Apigatewayv2)
+					if err != nil {
+						m.Logger().Errorf("could not get http and websocket apis output: %v", err)
+					}
+					if len(apiGatewayAPI) > 0 {
+						maps.Copy(infoapi, apiGatewayAPI)
+					}
+
+				}
+
```
In general, I try not to have specific AWS service API calls in the "generic" Fetch() function.

The Fetch() function handles metric collection for any metric namespace. I would try not to have conditional behavior based on a specific namespace ("AWS/ApiGateway" in this case) in the main code path.

gizas replied:
This is what `if strings.Contains(strings.ToLower(namespace), checkns_lower)`  does checks specific namespace == "AWS/ApiGateway"

gizas commented on metricbeat/docs/modules/aws.asciidoc:
```diff
@@ -146,6 +146,25 @@ Enforces the use of FIPS service endpoints. See <<aws-credentials-config,AWS cre
+      metrics:
+        - namespace: "AWS/ApiGateway"
+          resource_type: "apigateway:restapis"
+          limit_rest_api: 40
```
changed it to `max_apigateway_results` 
wdyt?

gizas replied:
apigateway_max_results is the last one

gizas commented:
I run another set of tests today the PR is fucntional (tested also   tags_filter)

![Screenshot 2024-10-02 at 3 14 27 PM](https://github.com/user-attachments/assets/4dbac3af-7f1b-4b19-8b78-0faaa630f98c)

Summarising:
- Renamed config option to apigateway_max_results
- Added checks for setting the correct limits of the apigateway_max_results
- The logic is inside the fetch function for now. We can consider breaking to another function/ file if we see that we have problems with other namespaces

Kavindu-Dodan reviewed (approved):
I only have this concern [1] otherwise looks okay. 

[1] - https://github.com/elastic/beats/pull/40755/files#r1784690644

cmacknz reviewed (approved):
Approving go.mod changes.

gizas commented on x-pack/metricbeat/module/aws/aws.go:
```diff
@@ -207,6 +208,18 @@ func NewMetricSet(base mb.BaseMetricSet) (*MetricSet, error) {
+	if config.LimitRestAPI != nil {
+		if *config.LimitRestAPI > 500 {
+			base.Logger().Debug("apigateway_max_results config value can not exceed value 500. Setting apigateway_max_results=25")
+			*config.LimitRestAPI = 25
```
You are correct. I misinterpreted initially the comment, now is set to 500

ycombinator commented:
Since this PR was merged, we're seeing the https://buildkite.com/elastic/beats-packaging-pipeline consistently failing like so when it tries to run `mage package` on Agentbeat:

```
# github.com/elastic/beats/v7/x-pack/agentbeat
/usr/local/go/pkg/tool/linux_amd64/link: running aarch64-linux-gnu-gcc failed: exit status 1
/usr/lib/gcc-cross/aarch64-linux-gnu/6/../../../../aarch64-linux-gnu/bin/ld.gold: internal error in maybe_apply_stub, at ../../gold/aarch64.cc:5407
collect2: error: ld returned 1 exit status
Error: running "go build -o build/golang-crossbuild/agentbeat-linux-arm64 -buildmode pie -gcflags=all=-N -l -tags=agentbeat -ldflags -X github.com/elastic/beats/v7/libbeat/version.buildTime=2024-10-03T12:37:52Z -X github.com/elastic/beats/v7/libbeat/version.commit=698951e8c895eff9d03a8d0ececadba3b8b4c6bb" failed with exit code 1
```

We're still investigating how to fix this forward but, if we can't figure something out soon, we may just need to revert this PR to make CI green again and buy us more time to figure out the problem.

cmacknz commented:
This PR makes it so that Beats depends on both AWS SDK v1 and AWS SDK v2.

```
	github.com/aws/aws-sdk-go-v2 v1.30.5
...
	github.com/aws/aws-sdk-go v1.54.19
```

cmacknz commented:
I suspect this is probably the source of the linker error it was introducing, hopefully it goes away if you only depend on `aws-sdk-go-v2` (and we should forbid bringing in the v1 SDK with a lint rule to avoid this in the future).

ycombinator commented:
I'm able to make the linker error go away if the code doesn't rely on `apigatewayv2.Client`: https://github.com/elastic/beats/pull/41129.  Not suggesting that this PR is the solution, but it may help narrow down the cause of the linker error.

cmacknz commented:
https://github.com/aws/aws-sdk-go is essentially deprecated and goes EOL on 7/31/2025, at which point we'll be forced to remove it as it will no longer receive security updates.

If one of them has to go it should be the v1 SDK.

cmacknz commented:
We ended up getting the same build error with a GCP dependency bump, so it's not AWS specifically causing this: https://github.com/elastic/beats/pull/41269
